File: Assets/Bundles/Shared/Resources/TS/Item/HeldItems/HeldItemManager.lua
-- Compiled with unity-ts v2.1.0-75
local Flamework = require("Shared/rbxts_include/node_modules/@easy-games/flamework-core/out/init").Flamework
local items = require("Shared/TS/Item/ItemDefinitions").items
local ItemType = require("Shared/TS/Item/ItemType").ItemType
local BlockHeldItem = require("Shared/TS/Item/HeldItems/BlockPlacement/BlockHeldItem").BlockHeldItem
local BreakBlockHeldItem = require("Shared/TS/Item/HeldItems/BlockPlacement/BreakBlockHeldItem").BreakBlockHeldItem
local MeleeHeldItem = require("Shared/TS/Item/HeldItems/Damagers/MeleeHeldItem").MeleeHeldItem
local HeldItem = require("Shared/TS/Item/HeldItems/HeldItem").HeldItem
local ProjectileLauncherHeldItem = require("Shared/TS/Item/HeldItems/ProjectileLauncher/ProjectileLauncherHeldItem").ProjectileLauncherHeldItem
local HeldItemState
do
	local _inverse = {}
	HeldItemState = setmetatable({}, {
		__index = _inverse,
	})
	HeldItemState.NONE = -1
	_inverse[-1] = "NONE"
	HeldItemState.CALL_TO_ACTION_START = 0
	_inverse[0] = "CALL_TO_ACTION_START"
	HeldItemState.CALL_TO_ACTION_END = 1
	_inverse[1] = "CALL_TO_ACTION_END"
	HeldItemState.ON_DESTROY = 2
	_inverse[2] = "ON_DESTROY"
end
-- One item manager per entity, calls functionality on currently equipped item for that entity
local HeldItemManager
do
	HeldItemManager = setmetatable({}, {
		__tostring = function()
			return "HeldItemManager"
		end,
	})
	HeldItemManager.__index = HeldItemManager
	function HeldItemManager.new(...)
		local self = setmetatable({}, HeldItemManager)
		return self:constructor(...) or self
	end
	function HeldItemManager:constructor(entity)
		self.heldItemMap = {}
		self.currentItemState = HeldItemState.NONE
		self.entity = entity
		self:Log("Creating Held Items")
		self.currentHeldItem = self:GetOrCreateHeldItem()
		-- Listen for item switches
		self.entity:GetInventory():ObserveHeldItem(function(itemStack)
			local _fn = self
			local _result = itemStack
			if _result ~= nil then
				_result = _result:GetMeta().displayName
			end
			_fn:Log("is equipping anew item: " .. tostring(_result))
			-- UnEquip last item
			if self.currentHeldItem ~= nil then
				self.currentHeldItem:OnUnEquip()
			end
			-- Equip the new item
			self.currentItemState = HeldItemState.NONE
			local _fn_1 = self
			local _result_1 = itemStack
			if _result_1 ~= nil then
				_result_1 = _result_1:GetMeta()
			end
			self.currentHeldItem = _fn_1:GetOrCreateHeldItem(_result_1)
			self.currentHeldItem:OnEquip()
		end)
	end
	function HeldItemManager:GetLabel()
		return self.entity.id
	end
	function HeldItemManager:Log(message)
		return nil
	end
	function HeldItemManager:GetOrCreateHeldItem(meta)
		if meta == nil then
			meta = items[ItemType.DEFAULT]
		end
		local _heldItemMap = self.heldItemMap
		local _itemType = meta.ItemType
		local item = _heldItemMap[_itemType]
		if item == nil then
			-- Create the held item instance
			local itemType = "NONE"
			self:Log("Creating Held Item...")
			if meta.melee then
				itemType = "MELEE"
				item = MeleeHeldItem.new(self.entity, meta)
			elseif meta.block then
				itemType = "BLOCK"
				item = BlockHeldItem.new(self.entity, meta)
			elseif meta.breakBlock then
				itemType = "BREAK BLOCK"
				item = BreakBlockHeldItem.new(self.entity, meta)
			elseif meta.ProjectileLauncher then
				itemType = "PROJECTILE LAUNCHER"
				item = ProjectileLauncherHeldItem.new(self.entity, meta)
			else
				warn("Entity " .. tostring(self.entity.id) .. " " .. meta.displayName .. " resorting to default held item logic")
				item = HeldItem.new(self.entity, meta)
			end
			self:Log("creating Held Item: " .. meta.displayName .. " of type: " .. itemType)
			local _heldItemMap_1 = self.heldItemMap
			local _itemType_1 = meta.ItemType
			local _item = item
			_heldItemMap_1[_itemType_1] = _item
		end
		return item
	end
	function HeldItemManager:TriggerNewState(itemState)
		-- Notify server of new State
		-- Dependency<LocalEntityController>().AddToMoveData("HeldItemState", itemState);
		(Flamework.resolveDependency("Bundles/Client/Controllers/Global/Character/LocalEntityController@LocalEntityController")):AddToMoveData("HeldItemState", {
			entityId = self.entity.id,
			state = itemState,
		})
		-- Network.ClientToServer.SetHeldItemState.Client.FireServer(this.entity.id, itemState);
		-- Handle the state locally
		self:OnNewState(itemState)
	end
	function HeldItemManager:OnNewState(itemState)
		self:Log("New State: " .. tostring(itemState))
		if self.currentItemState == itemState then
			return nil
		end
		if self.currentHeldItem == nil then
			error("Trying to interact without any held item!")
			return nil
		end
		self.currentItemState = itemState
		repeat
			local _fallthrough = false
			if itemState == (HeldItemState.CALL_TO_ACTION_START) then
				self.currentHeldItem:OnCallToActionStart()
				break
			end
			if itemState == (HeldItemState.CALL_TO_ACTION_END) then
				self.currentHeldItem:OnCallToActionEnd()
				break
			end
			if itemState == (HeldItemState.ON_DESTROY) then
				-- When destroyed un equip so any logic can clean itself up
				self.currentHeldItem:OnUnEquip()
				-- Fill current held item with default item
				self.currentHeldItem = self:GetOrCreateHeldItem()
			end
		until true
	end
end
return {
	HeldItemState = HeldItemState,
	HeldItemManager = HeldItemManager,
}
-- ----------------------------------
-- ----------------------------------

File: Assets/Typescript~/src/Shared/Item/HeldItems/HeldItemManager.ts
import { Dependency } from "@easy-games/flamework-core";
import { LocalEntityController } from "Client/Controllers/Global/Character/LocalEntityController";
import { CharacterEntity } from "../../Entity/Character/CharacterEntity";
import { items } from "../ItemDefinitions";
import { ItemMeta } from "../ItemMeta";
import { ItemType } from "../ItemType";
import { BlockHeldItem } from "./BlockPlacement/BlockHeldItem";
import { BreakBlockHeldItem } from "./BlockPlacement/BreakBlockHeldItem";
import { MeleeHeldItem } from "./Damagers/MeleeHeldItem";
import { HeldItem } from "./HeldItem";
import { ProjectileLauncherHeldItem } from "./ProjectileLauncher/ProjectileLauncherHeldItem";

export enum HeldItemState {
	NONE = -1,
	CALL_TO_ACTION_START = 0,
	CALL_TO_ACTION_END = 1,
	ON_DESTROY,
}

//One item manager per entity, calls functionality on currently equipped item for that entity
export class HeldItemManager {
	private entity: CharacterEntity;
	private heldItemMap = new Map<ItemType, HeldItem>();
	private currentHeldItem: HeldItem;
	private currentItemState: HeldItemState = HeldItemState.NONE;

	public GetLabel() {
		return this.entity.id;
	}

	private Log(message: string) {
		return;
		print("Entity " + this.entity.id + " " + message);
	}

	private GetOrCreateHeldItem(meta?: ItemMeta) {
		if (meta === undefined) {
			meta = items[ItemType.DEFAULT] as ItemMeta;
		}
		let item = this.heldItemMap.get(meta.ItemType);
		if (item === undefined) {
			//Create the held item instance
			let itemType = "NONE";
			this.Log("Creating Held Item...");
			if (meta.melee) {
				itemType = "MELEE";
				item = new MeleeHeldItem(this.entity, meta);
			} else if (meta.block) {
				itemType = "BLOCK";
				item = new BlockHeldItem(this.entity, meta);
			} else if (meta.breakBlock) {
				itemType = "BREAK BLOCK";
				item = new BreakBlockHeldItem(this.entity, meta);
			} else if (meta.ProjectileLauncher) {
				itemType = "PROJECTILE LAUNCHER";
				item = new ProjectileLauncherHeldItem(this.entity, meta);
			} else {
				warn("Entity " + this.entity.id + " " + meta.displayName + " resorting to default held item logic");
				item = new HeldItem(this.entity, meta);
			}
			this.Log("creating Held Item: " + meta.displayName + " of type: " + itemType);
			this.heldItemMap.set(meta.ItemType, item);
		}
		return item;
	}

	constructor(entity: CharacterEntity) {
		this.entity = entity;
		this.Log("Creating Held Items");
		this.currentHeldItem = this.GetOrCreateHeldItem();

		//Listen for item switches
		this.entity.GetInventory().ObserveHeldItem((itemStack) => {
			this.Log("is equipping anew item: " + itemStack?.GetMeta().displayName);
			//UnEquip last item
			if (this.currentHeldItem !== undefined) {
				this.currentHeldItem.OnUnEquip();
			}
			//Equip the new item
			this.currentItemState = HeldItemState.NONE;
			this.currentHeldItem = this.GetOrCreateHeldItem(itemStack?.GetMeta());
			this.currentHeldItem.OnEquip();
		});
	}

	//LOCAL CLIENT ONLY
	public TriggerNewState(itemState: HeldItemState) {
		//Notify server of new State
		// Dependency<LocalEntityController>().AddToMoveData("HeldItemState", itemState);
		Dependency<LocalEntityController>().AddToMoveData("HeldItemState", {
			entityId: this.entity.id,
			state: itemState,
		});
		// Network.ClientToServer.SetHeldItemState.Client.FireServer(this.entity.id, itemState);

		//Handle the state locally
		this.OnNewState(itemState);
	}

	public OnNewState(itemState: HeldItemState) {
		this.Log("New State: " + itemState);
		if (this.currentItemState === itemState) {
			return;
		}
		if (this.currentHeldItem === undefined) {
			error("Trying to interact without any held item!");
			return;
		}
		this.currentItemState = itemState;
		switch (itemState) {
			case HeldItemState.CALL_TO_ACTION_START:
				this.currentHeldItem.OnCallToActionStart();
				break;
			case HeldItemState.CALL_TO_ACTION_END:
				this.currentHeldItem.OnCallToActionEnd();
				break;
			case HeldItemState.ON_DESTROY:
				//When destroyed un equip so any logic can clean itself up
				this.currentHeldItem.OnUnEquip();
				//Fill current held item with default item
				this.currentHeldItem = this.GetOrCreateHeldItem();
		}
	}
}
